File: shared/downhill/ai.js
/**
 * Downhill Mayhem deterministic AI (integrate-multiplayer-downhill-mayhem-arcade
 * 3.2/3.4). Faithful extraction of the frozen source `aiThink`/`aiBoostWant`/
 * `predictAirRemaining`/`rampAheadFor` from
 * `games/downhill-mayhem/standalone.html`.
 *
 * The source used `Math.random()` for AI decisions; every one of those draws is
 * replaced by `ctx.rng()`, a deterministic per-match stream, so the authority
 * and its replay fixtures are reproducible. AI produce the same control shape a
 * human produces and are stepped by the same `stepRider` physics.
 *
 * `aiControl(course, rider, ctx)` returns a control object and may mutate the
 * rider's windup/cooldown state and emit strike events through `ctx.events`.
 */

import { clamp, FINISH_S } from './course.js';
import {
  G, DIFFS, TRICKS, PUNCH_S, PUNCH_LAT, PUNCH_DY,
  startTrick, tryStrike,
} from './rules.js';

export function predictAirRemaining(course, r) {
  let t = 0.3;
  for (let k = 0; k < 3; k++) {
    const h = r.y - course.heightAt(r.s + r.vs * t, r.lat);
    t = (r.vy + Math.sqrt(Math.max(0.01, r.vy * r.vy + 2 * G * Math.max(0.1, h)))) / G;
  }
  return t;
}

export function rampAheadFor(course, r) {
  for (const rp of course.ramps) {
    const d = rp.s0 - r.s;
    if (d > 3 && d < 32 && Math.abs((rp.latC || 0) - r.lat) < (rp.halfW || 6) + 2.5) return true;
  }
  return false;
}

export function aiBoostWant(course, r, maxV, cAhead, hunt) {
  const inp = r.inp;
  if (inp.brake || !r.grounded || r.vs >= maxV * 1.02 || r.meter <= 4) return false;
  const finalKick = r.s > FINISH_S - 400;
  if (r.boosting) return r.meter > (finalKick ? 4 : 8);
  if (finalKick) return true;
  let want = 999;
  if (hunt) want = Math.min(want, 12);
  if (rampAheadFor(course, r)) want = Math.min(want, 14);
  if (r.draftT > 0.3) want = Math.min(want, 17);
  if (r.rubber > 0.4) want = Math.min(want, 24);
  if (Math.abs(cAhead) < 0.02) want = Math.min(want, 30 + (r.slot % 3) * 7);
  if (r.meter > 76) want = Math.min(want, r.meter);
  return r.meter >= want;
}

/**
 * Generate one AI rider's control for this tick. `ctx` carries the field and
 * deterministic environment:
 *   { riders, difficulty, elapsed, reference, rng(), events, state, stats }
 */
export function aiControl(course, r, ctx) {
  const DIFF = DIFFS[ctx.difficulty] ?? DIFFS.mayhem;
  const riders = ctx.riders;
  const reference = ctx.reference ?? riders[0];
  const rng = ctx.rng ?? (() => 0.5);
  const dt = ctx.dt ?? (1 / 30);
  const elapsed = ctx.elapsed ?? 0;
  const control = { pedal: 0, brake: 0, steer: 0, hop: false, boost: false, punch: false, kick: false, trick: null };

  if (r.finished) { control.brake = 1; return control; }
  if (r.reactT > 0) { r.reactT -= dt; return control; }
  if (r.crashed) return control;

  // rubber-band vs the reference human.
  const gap = reference.s - r.s;
  r.rubber = gap >= 0 ? clamp(gap / DIFF.rubSat, 0, 1) * DIFF.rubP : clamp(gap / 160, -1, 0) * DIFF.leash;
  // comeback company: keep a struggling reference in sight.
  if (DIFF.company && gap < -70 && elapsed < 120 && (ctx.referenceCrashes ?? 0) < 6 && !r.finished) {
    const ahead = riders.filter((o) => o.isAI && !o.finished && o.s > reference.s);
    if (ahead.length === riders.filter((o) => o.isAI).length) {
      ahead.sort((a, b) => a.s - b.s);
      if (r === ahead[0] || r === ahead[1]) r.rubber = Math.min(r.rubber, -0.28);
    }
  }

  if (r.revengeT > 0) {
    r.revengeT -= dt;
    if (reference.finished || r.s > FINISH_S - 100) r.revengeT = 0;
  }
  const hunting = r.revengeT > 0;
  if (hunting && gap > 0) r.rubber = Math.max(r.rubber, Math.min(0.5 * DIFF.rev, 1.0));

  const look = 16 + r.vs * 0.9;
  const cAhead = course.sampleTrack(r.s + look).curv;
  const maxV = Math.sqrt((DIFF.corner + r.rubber * 2.6) * r.def.corner / Math.max(Math.abs(cAhead), 0.0004));
  control.brake = r.vs > maxV * 1.06 ? 1 : 0;
  control.pedal = control.brake ? 0 : 1;

  let target;
  if (hunting && DIFF.huntRace) {
    if (aiBoostWant(course, r, maxV, cAhead, true)) control.boost = true;
    target = reference.lat;
  } else if (hunting) {
    if (gap > 2.5) {
      const closeV = reference.vs + 1.5 * DIFF.rev + gap * 0.55;
      if (r.vs > closeV) { control.brake = 1; control.pedal = 0; }
      else if (gap > 3 && aiBoostWant(course, r, maxV, cAhead, true)) control.boost = true;
    } else if (gap < -3) { control.pedal = 0; control.brake = 1; }
    else {
      const want = gap > 1.2 ? reference.vs + Math.min(2.8 * DIFF.rev, Math.max(0.8, gap * 1.1)) : reference.vs;
      control.pedal = r.vs < want - 0.4 ? 1 : 0;
      control.brake = r.vs > want + 0.4 ? 1 : 0;
    }
    target = reference.lat;
  } else {
    target = clamp(-cAhead * 260, -3.3, 3.3) + r.lineBias
      + Math.sin(elapsed * r.wf + r.phase) * r.wamp * Math.max(0, 1 - r.rubber * 1.6);
    for (const o of riders) {
      if (o === r) continue;
      const ds = o.s - r.s;
      if (ds > 0.5 && ds < 7 && Math.abs(o.lat - r.lat) < 1.6) target += (r.lat >= o.lat ? 1.9 : -1.9);
    }
    for (const o of riders) {
      if (o === r || o.crashed || o.finished) continue;
      const ds = o.s - r.s;
      if (ds > -3 && ds < 0.5 && Math.abs(o.lat - r.lat) < 1.5 && (o.windupT >= 0 || (o.isHuman && r.punchCd > 0.4))) {
        target += (r.lat >= o.lat ? 2.2 : -2.2);
      }
    }
    if (aiBoostWant(course, r, maxV, cAhead, false)) control.boost = true;
  }

  // obstacle avoidance.
  const b0 = Math.floor(r.s / 10), obs = [];
  for (let bi = b0; bi <= b0 + 4; bi++) {
    const list = course.colliderBuckets.get(bi);
    if (!list) continue;
    for (const t of list) { const ds = t.s - r.s; if (ds > 2 && ds < 34) obs.push(t); }
  }
  if (obs.length) {
    const blockers = obs.filter((t) => Math.abs(t.lat - target) < 1.55);
    if (blockers.length) {
      let lo = 99, hi = -99;
      for (const t of blockers) { lo = Math.min(lo, t.lat); hi = Math.max(hi, t.lat); }
      const clearOf = (lat) => obs.every((t) => Math.abs(t.lat - lat) >= 1.55);
      let cand = [lo - 2.1, hi + 2.1].filter((c) => clearOf(c) && Math.abs(c) <= 7.2);
      if (!cand.length) {
        let glo = 99, ghi = -99;
        for (const t of obs) { glo = Math.min(glo, t.lat); ghi = Math.max(ghi, t.lat); }
        cand = [glo - 2.1, ghi + 2.1];
      }
      target = cand.reduce((a, b) => Math.abs(b - r.lat) < Math.abs(a - r.lat) ? b : a);
      if (!hunting && r.vs > 18 + DIFF.corner * 0.9) { control.brake = true; control.boost = false; }
    }
  }
  target = clamp(target, -7.2, 7.2);
  const desiredVlat = clamp((target - r.lat) * (hunting ? 2.4 : 1.8), -6, 6);
  control.steer = clamp((desiredVlat - r.vlat) * 0.45, -1, 1);

  if (r.grounded) {
    r.trickRolled = false;
    const rampH = course.rampHeightAt(r.s, r.lat);
    if (rampH > 0.05 && r.def.trick > 0.4 && !hunting) {
      let onLip = false;
      for (const rp of course.ramps) {
        if (r.s >= rp.s0 && r.s <= rp.s0 + rp.len && (r.s - rp.s0) / rp.len > 0.72) { onLip = true; break; }
      }
      if (onLip) control.hop = rng() < dt * (1.2 + r.def.trick * 2.4);
    }
  } else if (!r.trick && !r.trickRolled && r.airTime > 0.24 && !hunting) {
    r.trickRolled = true;
    const rem = predictAirRemaining(course, r) - (0.12 - r.def.crashy * 0.05);
    if (rem > TRICKS.nohander.dur && rng() < 0.36 + r.def.trick * 0.58) {
      const menu = r.def.trick > 0.7 ? ['heel', 'backflip', 'superman', 'nohander']
        : r.def.trick > 0.45 ? ['superman', 'backflip', 'nohander']
          : ['nohander', 'superman'];
      let pick = null;
      for (const k of menu) if (TRICKS[k].dur < rem) { pick = k; break; }
      if (pick) startTrick(r, pick);
    }
  }

  // punching/kicking with windup tell.
  if (!ctx.state) ctx.state = { lastPunchOnHumanAt: -99 };
  const state = ctx.state;
  if (r.windupT >= 0) {
    r.windupT += dt;
    if (r.windupT >= (r.windupDur || 0.45)) {
      r.windupT = -1;
      const t = r.windupTarget;
      if (t && !t.crashed && t.invuln <= 0 && Math.abs(t.s - r.s) < PUNCH_S && Math.abs(t.lat - r.lat) < PUNCH_LAT) {
        tryStrike(r, r.strikeKind, riders, ctx, ctx.events);
      }
      r.punchCd = (r.revengeT > 0 ? 0.7 + rng() * 0.5 : r.grudge ? 2.5 + rng() * 2 : 4 + rng() * 3) * DIFF.cd;
    }
  } else if (r.punchCd <= 0 && r.grounded) {
    if (hunting) {
      const dsP = Math.abs(reference.s - r.s), dlP = Math.abs(reference.lat - r.lat);
      if (dsP < 2.0 && dlP < 1.3 && Math.abs(reference.y - r.y) < PUNCH_DY
        && !reference.crashed && reference.invuln <= 0 && !reference.finished) {
        r.windupT = 0; r.windupTarget = reference; r.strikeKind = rng() < 0.5 ? 'kick' : 'punch';
        if (ctx.stats) ctx.stats[r.strikeKind === 'kick' ? 'aiKick' : 'aiPunch']++;
        r.windupDur = Math.max(0.26, 0.45 / DIFF.rev);
        r.strikeSide = Math.sign(reference.lat - r.lat) || 1;
      }
    } else {
      for (const o of riders) {
        if (o === r || o.crashed || o.invuln > 0 || o.finished) continue;
        const ds = Math.abs(o.s - r.s), dl = Math.abs(o.lat - r.lat);
        if (ds < 2.0 && dl < 1.3 && Math.abs(o.y - r.y) < PUNCH_DY) {
          const isHuman = o.isHuman;
          if (isHuman && elapsed - state.lastPunchOnHumanAt < 2) continue;
          if (rng() < r.def.aggr * (isHuman ? (r.grudge ? 4 : 2.6) : 0.45 * (1 - r.rubber)) * DIFF.aggr * dt * 2.2) {
            r.windupT = 0; r.windupTarget = o; r.strikeKind = rng() < 0.45 ? 'kick' : 'punch';
            if (ctx.stats) ctx.stats[r.strikeKind === 'kick' ? 'aiKick' : 'aiPunch']++;
            r.windupDur = 0.45;
            r.strikeSide = Math.sign(o.lat - r.lat) || 1;
            if (isHuman) state.lastPunchOnHumanAt = elapsed;
            break;
          }
        }
      }
    }
  }
  return control;
}

File: server_elixir/lib/afterlight/activities/downhill_mayhem/ai.ex
defmodule Afterlight.Activities.DownhillMayhem.AI do
  @moduledoc """
  Server-owned Downhill Mayhem AI
  (integrate-multiplayer-downhill-mayhem-arcade 8.3).

  Server-only control generation: a racing line with corner braking, obstacle
  avoidance, situational boost, ramp hops/tricks, rubber-banding against a
  reference human, revenge hunting after being decked, and combat expressed as
  `punch`/`kick` control edges (the shared physics resolves the actual hit).
  Randomness uses the BEAM process-local `:rand` state, seeded deterministically
  by the session, so a match is reproducible.
  """

  alias Afterlight.Activities.DownhillMayhem
  alias Afterlight.Activities.DownhillMayhem.Course

  @g 11.5
  @finish_s 2300.0

  defp clamp(v, a, b), do: if(v < a, do: a, else: if(v > b, do: b, else: v))
  defp rnd, do: :rand.uniform()
  defp sign(v), do: if(v < 0, do: -1, else: if(v > 0, do: 1, else: 0))

  def predict_air_remaining(course, r) do
    Enum.reduce(0..2, 0.3, fn _k, t ->
      h = r.y - Course.height_at(course, r.s + r.vs * t, r.lat)
      (r.vy + :math.sqrt(max(0.01, r.vy * r.vy + 2 * @g * max(0.1, h)))) / @g
    end)
  end

  def ramp_ahead_for(course, r) do
    Enum.any?(course.ramps, fn rp ->
      d = rp["s0"] - r.s
      d > 3 and d < 32 and abs((rp["latC"] || 0) - r.lat) < (rp["halfW"] || 6) + 2.5
    end)
  end

  defp ai_boost_want(course, r, max_v, c_ahead, hunt) do
    cond do
      r.inp.brake == 1 or not r.grounded or r.vs >= max_v * 1.02 or r.meter <= 4 -> false
      r.boosting -> r.meter > (if r.s > @finish_s - 400, do: 4, else: 8)
      r.s > @finish_s - 400 -> true
      true ->
        want =
          Enum.min([
            if(hunt, do: 12, else: 999),
            if(ramp_ahead_for(course, r), do: 14, else: 999),
            if(r.draft_t > 0.3, do: 17, else: 999),
            if(r.rubber > 0.4, do: 24, else: 999),
            if(abs(c_ahead) < 0.02, do: 30 + rem(r.slot, 3) * 7, else: 999),
            if(r.meter > 76, do: r.meter, else: 999)
          ])

        r.meter >= want
    end
  end

  @doc "Return a control map for one AI rider this tick."
  def control(course, r, ctx) do
    diff = DownhillMayhem.diff(Map.get(ctx, :difficulty, "mayhem"))
    riders = Map.values(Map.get(ctx, :riders, %{}))
    reference = Map.get(ctx, :reference) || List.first(riders) || r
    dt = Map.get(ctx, :dt, 1 / 30)
    elapsed = Map.get(ctx, :elapsed, 0.0)
    control = DownhillMayhem.neutral_controls()

    cond do
      r.finished -> %{control | brake: 1}
      r.react_t > 0 -> control
      r.crashed -> control
      true -> think(course, r, ctx, diff, riders, reference, dt, elapsed)
    end
  end

  defp think(course, r, _ctx, diff, riders, reference, dt, elapsed) do
    gap = reference.s - r.s
    rubber = if gap >= 0, do: clamp(gap / diff.rub_sat, 0.0, 1.0) * diff.rub_p, else: clamp(gap / 160, -1.0, 0.0) * diff.leash
    r = %{r | rubber: rubber}

    r =
      if r.revenge_t > 0 do
        if reference.finished or r.s > @finish_s - 100, do: %{r | revenge_t: 0.0}, else: %{r | revenge_t: r.revenge_t - dt}
      else
        r
      end

    hunting = r.revenge_t > 0
    r = if hunting and gap > 0, do: %{r | rubber: max(r.rubber, min(0.5 * diff.rev, 1.0))}, else: r

    look = 16 + r.vs * 0.9
    c_ahead = Course.sample_track(course, r.s + look).curv
    max_v = :math.sqrt((diff.corner + r.rubber * 2.6) * r.def.corner / max(abs(c_ahead), 0.0004))

    braking = r.vs > max_v * 1.06
    control = %{DownhillMayhem.neutral_controls() | brake: if(braking, do: 1, else: 0), pedal: if(braking, do: 0, else: 1)}

    target =
      cond do
        hunting -> reference.lat
        true ->
          base = clamp(-c_ahead * 260, -3.3, 3.3) + r.line_bias + :math.sin(elapsed * r.wf + r.phase) * r.wamp * max(0, 1 - r.rubber * 1.6)

          base =
            Enum.reduce(riders, base, fn o, acc ->
              ds = o.s - r.s
              cond do
                o.slot == r.slot -> acc
                ds > 0.5 and ds < 7 and abs(o.lat - r.lat) < 1.6 -> acc + (if r.lat >= o.lat, do: 1.9, else: -1.9)
                ds > -3 and ds < 0.5 and abs(o.lat - r.lat) < 1.5 and (o.windup_t >= 0 or (o.is_human and r.punch_cd > 0.4)) -> acc + (if r.lat >= o.lat, do: 2.2, else: -2.2)
                true -> acc
              end
            end)

          base
      end

    target = avoid_obstacles(course, r, target)
    target = clamp(target, -7.2, 7.2)
    desired_vlat = clamp((target - r.lat) * (if hunting, do: 2.4, else: 1.8), -6.0, 6.0)
    control = %{control | steer: clamp((desired_vlat - r.vlat) * 0.45, -1.0, 1.0)}

    control =
      if Map.get(diff, :hunt_race, false) and hunting do
        if ai_boost_want(course, r, max_v, c_ahead, true), do: %{control | boost: true}, else: control
      else
        if ai_boost_want(course, r, max_v, c_ahead, hunting), do: %{control | boost: true}, else: control
      end

    {r, control} = jump_and_tricks(course, r, control, hunting, dt)
    combat(r, control, diff, riders, reference, elapsed)
  end

  defp avoid_obstacles(course, r, target) do
    b0 = trunc(Float.floor(r.s / 10))

    obs =
      Enum.flat_map(b0..(b0 + 4), fn bi -> Map.get(course.collider_buckets, bi, []) end)
      |> Enum.filter(&(&1["s"] - r.s > 2 and &1["s"] - r.s < 34))

    blockers = Enum.filter(obs, &(abs(&1["lat"] - target) < 1.55))

    if obs != [] and blockers != [] do
      lo = blockers |> Enum.map(& &1["lat"]) |> Enum.min()
      hi = blockers |> Enum.map(& &1["lat"]) |> Enum.max()
      clear_of = fn lat -> Enum.all?(obs, &(abs(&1["lat"] - lat) >= 1.55)) end
      cand = Enum.filter([lo - 2.1, hi + 2.1], fn c -> clear_of.(c) and abs(c) <= 7.2 end)

      cand =
        if cand == [] do
          [Enum.min(Enum.map(obs, & &1["lat"])) - 2.1, Enum.max(Enum.map(obs, & &1["lat"])) + 2.1]
        else
          cand
        end

      Enum.min_by(cand, &abs(&1 - r.lat))
    else
      target
    end
  end

  defp jump_and_tricks(course, r, control, hunting, dt) do
    cond do
      r.grounded ->
        r = %{r | trick_rolled: false}
        ramph = Course.ramp_height_at(course, r.s, r.lat)

        if ramph > 0.05 and r.def.trick > 0.4 and not hunting do
          on_lip =
            Enum.any?(course.ramps, fn rp ->
              r.s >= rp["s0"] and r.s <= rp["s0"] + rp["len"] and (r.s - rp["s0"]) / rp["len"] > 0.72
            end)

          if on_lip, do: {r, %{control | hop: rnd() < dt * (1.2 + r.def.trick * 2.4)}}, else: {r, control}
        else
          {r, control}
        end

      r.trick == nil and not r.trick_rolled and r.air_time > 0.24 and not hunting ->
        r = %{r | trick_rolled: true}
        rem_t = predict_air_remaining(course, r) - (0.12 - r.def.crashy * 0.05)
        tricks = DownhillMayhem.tricks()

        if rem_t > tricks["nohander"].dur and rnd() < 0.36 + r.def.trick * 0.58 do
          menu =
            cond do
              r.def.trick > 0.7 -> ["heel", "backflip", "superman", "nohander"]
              r.def.trick > 0.45 -> ["superman", "backflip", "nohander"]
              true -> ["nohander", "superman"]
            end

          pick = Enum.find(menu, fn k -> tricks[k].dur < rem_t end)
          if pick, do: {%{r | trick: pick, trick_t: 0.0}, control}, else: {r, control}
        else
          {r, control}
        end

      true ->
        {r, control}
    end
  end

  # Combat as control edges: the shared physics performs the strike, cooldown and
  # grudge/revenge bookkeeping, so AI and humans share one resolution path.
  defp combat(r, control, diff, riders, reference, elapsed) do
    cond do
      r.punch_cd > 0 -> control
      not r.grounded -> control
      r.revenge_t > 0 ->
        if in_reach?(reference, r) and not reference.crashed and reference.invuln <= 0 and not reference.finished do
          kind = if rnd() < 0.5, do: :kick, else: :punch
          Map.put(control, kind, true)
        else
          control
        end

      true ->
        target =
          Enum.find(riders, fn o ->
            o.slot != r.slot and not o.crashed and o.invuln <= 0 and not o.finished and in_reach?(o, r)
          end)

        roll = r.def.aggr * (if target, do: (if target.is_human, do: (if r.grudge, do: 4, else: 2.6), else: 0.45 * (1 - r.rubber)), else: 0) * diff.aggr * 0.12

        if target != nil and :rand.uniform() < roll and elapsed > 0 do
          kind = if rnd() < 0.45, do: :kick, else: :punch
          Map.put(control, kind, true)
        else
          control
        end
    end
  end

  defp in_reach?(o, r) do
    abs(o.s - r.s) < 2.0 and abs(o.lat - r.lat) < 1.3 and abs(o.y - r.y) < 1.4
  end
end
